docs/params.rst

```diff
@@ -0,0 +1,148 @@
+    expect or support.
+
+-   Since arguments can come from the command line, and those are always
+    strings, you have to write code to convert arguments to a different type
+    when needed.
+
+-   If you want argument validation, such as making sure that arguments are of
+    the right type, or that required arguments are present, you must implement
+    validation yourself.
+
+scrapy-spider-metadata allows overcoming those limitations.
+
+.. _define-params:
+
+Defining supported parameters
+=============================
+
+To define spider parameters, define a subclass of `pydantic.BaseModel`_,
+and then point to it from the ``"params"`` key of a :class:`dict` ``meta``
+class variable of your spider:
+
+.. _pydantic.BaseModel: https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel
+
+.. code-block:: python
+
+    from pydantic import BaseModel
+    from scrapy import Spider
+
+    class MyParams(BaseModel):
+        foo: int
+
+    class MySpider(Spider):
+        name = "my_spider"
+        meta = {
+            "params": MyParams,
+        }
+
+To learn how to define parameters in your `pydantic.BaseModel`_
+subclass, see the `Pydantic usage documentation
+<https://docs.pydantic.dev/latest/usage/models/>`_.
+
+.. _define-params-limits:
+
+However, scrapy-spider-metadata only supports a subset of Pydantic features:
+
+-   Only the following parameter types are supported: :class:`str`,
+    :class:`int`, :class:`float`, :class:`bool`, :class:`~enum.Enum`,
+    :class:`~enum.IntEnum`.
+
+-   Defining multiple types for a given parameter is not supported.
+
+Using other Pydantic features can prevent scrapy-spider-metadata from working
+as expected.
+
+.. note:: If you wish to use a Pydantic feature not listed above, please see if
+    it has already been requests in `our issue tracker
+    <https://github.com/scrapy-plugins/scrapy-spider-metadata/issues>`_, and if
+    not, feel free to request it.
+
+
+Converting and validating arguments
+===================================
+
+Add the :class:`~scrapy_spider_metadata.ParamSpiderMixin` mixin to a spider
+with :ref:`defined parameters <define-params>` to make the spider:
+
+-   Halt with an exception if there are missing arguments or any provided
+    argument does not match the defined parameter validation rules.
+
+-   Convert arguments to the expected type.
+
+For example, if you modify the :ref:`example above <define-params>` to use the
+mixin:
+
+.. code-block:: python
+    :emphasize-lines: 3, 8
+
+    from pydantic import BaseModel
+    from scrapy import Spider
+    from scrapy_spider_metadata import ParamSpiderMixin
+
+    class MyParams(BaseModel):
+        foo: int
+
+    class MySpider(ParamSpiderMixin, Spider):
+        name = "my_spider"
+        meta = {
+            "params": MyParams,
+        }
+
+And then run the spider with the ``foo`` parameter set to the string ``"42"``,
+``self.foo`` in your spider will be the :class:`int` value ``42``, instead of
+``"42"``.
```

Comment:
In a previous proposal the idea was to have a special attribute, with parsed parameters, which gets somehow populated based on the arguments passed to the spider. I kind-of liked it for few reasons:

1. This attribute was typed (it'd be of type MyParams in this example), which means better IDE support and an option to catch errors via mypy for individual parameters (like int foo in this example). It seems it might be possible to do the typing in a way the attribute type is handled automatically, based on the type set in meta.
2. We're moving away from attributes "polluting" the top level of the spider instance.
3. Probably not important, but it was possible to access both raw and parsed parameters. It was also possible to have argument parsing lazy - again, not sure about the exact use case.

On the other hand, approach here seems to 

1. Be less verbose for the user.
2. Follow more closely the current Scrapy approach of using arguments as attributes; probably, less changes in existing spiders would be required to move.

Any other pro/cons you see here @Gallaecio?

Reply:
It was like this in the previous proposal:

```py
class MySpider(ParamSpiderMixin, Spider):
    name = "my_spider"
    meta = {
        "params": MyParams,
    }
    params: MyParams

# or

class MySpider(ParamSpiderMixin, Spider):
    name = "my_spider"
    class Meta:
        params_schema: MyParams
    params: MyParams

# it seems the following syntax is also possible, if we fancy it

class MySpider(ParamSpiderMixin, Spider):
    name = "my_spider"
    class Meta:
        params_schema: MyParams
      # .params is an attribute of ParamSpideMixin

# or, somewhat similar:
class MySpider(ParamSpiderMixin[MyParams], Spider):
    name = "my_spider"

    # params: MyParams is an attribute of ParamSpideMixin
    # the mixin adds to spider meta, maybe the jsonschema directly

```

Reply:
Heh, it sound a bit cryptic to me :) 

> past tense: parameterized; past participle: parameterized

    describe or represent in terms of a parameter or parameters.
    "the nuclear charge distribution can be parameterized directly using a suitable mathematical form"

Some other ideas: 

* ParseArguments[MyParams]
* ParseArgumentsMixin[MyParams]
* ArgumentSchema[MyParams]
* AcceptsArguments[MyParams]
* Args[MyParams]
* SpiderArguments[MyParams]
* SpiderArgumentsSchema[MyParams]
* ParsesArguments[MyParams]
* ...

Reply:
I usually prefer longer, more-descriptive names, but since this is something to be included in usually-long lines (spider class definitions) and there is a beautiful symmetry with inheriting `Args` causing `self.args` to be provided, I went for that one :slightly_smiling_face: